### PR TITLE
cheribsdtest: infrastructure to spawn children

### PR DESCRIPTION
**Pull request by brooksdavis**
Add infrastructure to spawn children which exec and run a per-test function.

Supersedes #1802

### CODE REVIEW

**Comment by brooksdavis** [comment]
I've added some tests partially extracted from https://github.com/brianli-sri/cheribsd/commit/6cb113cdda01ea31260a1ade6c1e4c59bee3dddf

**Comment by brooksdavis** [comment]
One downside of the tests I've added is that they add noticeably to the runtime of cheribsdtest. I'm not sure the _once vs _split distinction is actually useful in these cases and that would cut the added runtime in half.

File: bin/cheribsdtest/cheribsdtest.c [hunk 1]
```diff
@@ -620,6 +625,102 @@ cheribsdtest_run_test_name(const char *name)
...
+	pid_t pid;
+
+	exec_args = mk_exec_args(running_test);
+
+	switch (mode) {
+	case SC_MODE_POSIX_SPAWN:
+		error = posix_spawn(&pid, exec_args[0], NULL, NULL, exec_args,
+		    environ);
+		if (error != 0)
+			pid = -1;
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest.c:700]: Does this need to set `errno` to `error`?

File: bin/cheribsdtest/cheribsdtest_vm.c [hunk 2]
```diff
@@ -1941,6 +1942,265 @@ CHERIBSDTEST(cheri_revoke_lib_fork_split,
...
+
+CHERIBSDTEST(cheri_revoke_lib_fork_exec_once_prior,
+    "revoke in a fork+exec'd child after revoking once",
+    .ct_child_func = cheri_revoke_lib_child_once)
+{
+	cheri_revoke_lib_spawn_common(SC_MODE_FORK, TCLR_MODE_LOAD_ONCE);
+}
+
+CHERIBSDTEST(cheri_revoke_lib_fork_exec_split_prior,
+    "split revoke in a fork+exec'd child after revokign once",
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_vm.c:2049]: s/revokign/revoking/ here and in a few other places

**Review by bsdjhb — COMMENTED** [review]
I don't know that we need the full set of permutations for the revoke tests vs using a subset of tests that cover all of the orthogonal spokes (e.g. at least one test of each spawn method).  You might want all of the revoke type x pre-fork revoke, but perhaps spread those six across the spawn types?  FWIW, I think `rfork()` is quite rare in practice, and `posix_spawn()` is just `vfork()` + `execve()` in libc, so it's probably most critical to test `fork()` and `vfork()` / `posix_spawn()`.

**Comment by brooksdavis** [comment]
> I don't know that we need the full set of permutations for the revoke tests vs using a subset of tests that cover all of the orthogonal spokes (e.g. at least one test of each spawn method). You might want all of the revoke type x pre-fork revoke, but perhaps spread those six across the spawn types? FWIW, I think `rfork()` is quite rare in practice, and `posix_spawn()` is just `vfork()` + `execve()` in libc, so it's probably most critical to test `fork()` and `vfork()` / `posix_spawn()`.

I've cut it down from 24 to 10 tests. I've got a single bare rfork() test and these days posix_spawn uses rfork(RFSPAWN) and  falls back to vfork() only if that fails. I think it's important to test the different pre-spawn states across multiple methods because I hit different bugs from different cases around that in caprevoke.

File: bin/cheribsdtest/cheribsdtest_vm.c [hunk 3]
```diff
@@ -1933,6 +1934,182 @@ CHERIBSDTEST(cheri_revoke_lib_fork_split,
...
+ * cheri_revoke_lib_child_* - test that execed children can revoke
+ *
+ * We selectively test along three axes:
+ *   spawn method: fork+execve, rfork+execve, vfork+execve, posix_spawn
+ *   pre-fork revoke: none, once, opened
+ *   revoke type: once, split
+ *
+ * Testing all 24 cases is seems excessive so we limit rfork+execve to a
+ * single test (posix_spawn being built on rfork) and alternate between
+ * once and split as the different has not previously resulted in bugs.
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_vm.c:1947]: s/different/difference/

**Review by bsdjhb — APPROVED** [review]

**Pull request merged**